internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,195 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver implements the default resolver that creates child
+// resolvers to resolver targetURI as well as proxy adress.
+
+package delegatingresolver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+type delegatingResolver struct {
+	target         resolver.Target
+	cc             resolver.ClientConn
+	isProxy        bool
+	targetResolver resolver.Resolver
+	ProxyResolver  resolver.Resolver
+	targetAddrs    []resolver.Address
+	proxyAddrs     []resolver.Address
+	proxyURL       *url.URL
+	mu             sync.Mutex // Protects the state below
+
+}
+
+type innerClientConn struct {
+	*delegatingResolver
+	resolverType string
+}
+
+var (
+	// The following variable will be overwritten in the tests.
+	httpProxyFromEnvironment = http.ProxyFromEnvironment
+)
+
+func mapAddress(address string) (*url.URL, error) {
+	req := &http.Request{
+		URL: &url.URL{
+			Scheme: "https",
+			Host:   address,
+		},
+	}
+	url, err := httpProxyFromEnvironment(req)
+	if err != nil {
+		return nil, err
+	}
+	return url, nil
+}
+
+func New(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions, targetResolverBuilder resolver.Builder) (resolver.Resolver, error) {
+	r := &delegatingResolver{
+		target: target,
+		cc:     cc,
+	}
+	var err error
+	r.proxyURL, err = mapAddress(target.Endpoint())
+	if err != nil {
+		return nil, err
+	}
+	r.isProxy = true
+	if r.proxyURL == nil {
+		r.isProxy = false
+	}
+	if !r.isProxy {
+		return targetResolverBuilder.Build(target, cc, opts)
+	}
+	if target.URL.Scheme == "dns" {
+		r.targetAddrs = []resolver.Address{{Addr: r.target.Endpoint()}}
+	} else {
+		r.targetResolver, err = targetURIResolver(target, opts, targetResolverBuilder, r)
+	}
+	r.ProxyResolver, err = proxyURIResolver(r.proxyURL, opts, r)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
```

Comment:
I feel like this could be simplified by reorganizing the code. How about something like

```golang
	proxyURL, err := mapAddress(target.Endpoint())
	if err != nil {
		// actually add some useful information to the error returned from mapAddress before returning from here
		return nil, err 
	}
	if proxyURL == nil {
		// Add an info log here.
		return targetResolverBuilder.Build(target, cc, opts)
	}
	
	// Add a comment here about why we don't create a target resolver if scheme is `dns`.
	var targetAdds []resolver.Address
	var targetResolver *resolver.Resolver
	if target.URL.Scheme == "dns" {
		targetAddrs = []resolver.Address{{Addr: r.target.Endpoint()}}
	} else {
		// Specify in the docstring that targetResolverBuilder should not be nil. 
		// If so, the channel should error out and not call this function.
		var err error
		targetResolver, err = targetResolverBuilder.Build(...)
		if err != nil {
			// return error with some extra information
		}
	}
	
	r.ProxyResolver, err = proxyURIResolver(r.proxyURL, opts, r)
	if err != nil {
		return nil, err
	}

	return &delegatingResolver{
		target: target,
		cc:     cc,
		proxyURL: proxyURL
		targetAddrs: targetAdds,
		targetResolver: targetResolver,
		proxyResolver: proxyResolver
	}, nil
```